### ISSUE TITLE
bitwise shift operations on DPU

### ISSUE THREAD

**Issue by dlagneux-upmem** [upmem/llvm-project#57]
ETH reported invalid results when doing bitwise shift operations on the DPU. Here is a simplified example:

```
#include <iso646.h>
#include <stdio.h>

int main() {
  printf("variable inside a loop \n");
  for (unsigned int i = 1; i < 5; i++) {
    printf("%d- %llx \n", i,~(1ULL << (32-i)));
  }

  printf("variable without  a loop \n");

  unsigned int x = 1;
  printf("%d- %llx \n",x, ~(1ULL << (32-x)));
  x = 2;
  printf("%d- %llx \n",x, ~(1ULL << (32-x)));
  x = 3;
  printf("%d- %llx \n",x, ~(1ULL << (32-x)));
  x = 4;
  printf("%d- %llx \n",x, ~(1ULL << (32-x)));

  printf("no variable\n");

  printf("%d- %llx \n",1, ~(1ULL << (32-1)));
  printf("%d- %llx \n",2, ~(1ULL << (32-2)));
  printf("%d- %llx \n",3, ~(1ULL << (32-3)));
  printf("%d- %llx \n",4, ~(1ULL << (32-4)));

  return 0;
}
```
Effective:
```
variable inside a loop
1- ffffffff7fffffff
2- ffffffff3fffffff
3- ffffffff1fffffff
4- ffffffff0fffffff
variable without  a loop
1- ffffffff7fffffff
2- ffffffff3fffffff
3- ffffffff1fffffff
4- ffffffff0fffffff
no variable
1- ffffffff7fffffff
2- ffffffffbfffffff
3- ffffffffdfffffff
4- ffffffffefffffff
```

Expected:
```
variable inside a loop
1- ffffffff7fffffff
2- ffffffffbfffffff
3- ffffffffdfffffff
4- ffffffffefffffff
variable without  a loop
1- ffffffff7fffffff
2- ffffffffbfffffff
3- ffffffffdfffffff
4- ffffffffefffffff
no variable
1- ffffffff7fffffff
2- ffffffffbfffffff
3- ffffffffdfffffff
4- ffffffffefffffff
```

We tried to understand the behavior using dpu.dev: https://dpu.dev/z/Tvm6QW, but the resulting  assembly code seems weird.
@JordiChauzi Can you please develop the "weird" behavior?

**Comment by dlagneux-upmem** [comment]
We still have an issue when compiling with -O0

**Comment by JordiChauzi** [comment]
Closed in favor of #58
